src/incremental/serialize.ml

```diff
@@ -38,22 +40,35 @@ let type_to_file_name = function
   | CilFile -> cil_file_name
   | VersionData -> version_map_filename
 
+let solver_data : Obj.t option ref = ref None
+
 (** Loads data for incremental runs from the appropriate file *)
 let load_data (data_type: incremental_data_kind) =
-  let p = Filename.concat (gob_results_dir ()) (type_to_file_name data_type) in
-  unmarshal p
+  if server () then
+    match data_type with
+    | SolverData -> !solver_data |> Option.get |> Obj.obj
+    | _ -> raise (Invalid_argument "Can only load solver data")
+  else
+    let p = Filename.concat (gob_results_dir ()) (type_to_file_name data_type) in
+    unmarshal p
 
 (** Stores data for future incremental runs at the appropriate file, given the data and what kind of data it is. *)
-let store_data (data: 'a) (data_type: incremental_data_kind) =
-  ignore @@ Goblintutil.create_dir (gob_directory ());
-  let d = gob_results_tmp_dir () in
-  ignore @@ Goblintutil.create_dir d;
-  let p = Filename.concat (d) (type_to_file_name data_type) in
-  marshal data p
+let store_data (data : 'a) (data_type : incremental_data_kind) =
+  if server () then
+    match data_type with
+    | SolverData -> solver_data := Some (Obj.repr data)
+    | _ -> ()
+  else (
+    ignore @@ Goblintutil.create_dir (gob_directory ());
+    let d = gob_results_tmp_dir () in
+    ignore @@ Goblintutil.create_dir d;
+    let p = Filename.concat d (type_to_file_name data_type) in
+    marshal data p)
 
 (** Deletes previous analysis results and moves the freshly created results there.*)
 let move_tmp_results_to_results () =
-  if Sys.file_exists (gob_results_dir ()) then begin
-    Goblintutil.rm_rf (gob_results_dir ());
-  end;
-  Sys.rename (gob_results_tmp_dir ()) (gob_results_dir ())
+  if not (server ()) then (
```

Comment:
Would we not still need to do this?

Reply:
Since I hijack all calls to `store_data`, we never actually create the `gob_results_tmp_dir ()` directory in the server mode, so the `Sys.rename` throws.